### PR TITLE
Rewrite ReactPy-Router

### PR DESCRIPTION
**Pull request by Archmonger**
## Description

Fixes a boatload of rendering bugs and aligns the API with react router.

### Changed

-   Bump GitHub workflows
-   Rename `use_query` to `use_search_params`.
-   Rename `simple.router` to `browser_router`.
-   Rename `SimpleResolver` to `StarletteResolver`.
-   Rename `CONVERSION_TYPES` to `CONVERTERS`.
-   Change "Match Any" syntax from a star `*` to `{name:any}`.
-   Rewrite `reactpy_router.link` to be a server-side component.
-   Simplified top-level exports within `reactpy_router`.

### Added

-   New error for ReactPy router elements being used outside router context.
-   Configurable/inheritable `Resolver` base class.
-   Add debug log message for when there are no router matches.
-   Add slug as a supported type.

### Fixed

-   Fix bug where changing routes could cause render failure due to key identity.
-   Fix bug where "Match Any" pattern wouldn't work when used in complex or nested paths.
-   Fix bug where `link` elements could not have `@component` type children.
-   Fix bug where the ReactPy would not detect the current URL after a reconnection.
-   Fixed flakey tests on GitHub CI by adding click delays.

## Checklist

Please update this checklist as you complete each item:

-   [x] Tests have been developed for bug fixes or new functionality.
-   [x] The changelog has been updated, if necessary.
-   [x] Documentation has been updated, if necessary.
-   [x] GitHub Issues closed by this PR have been linked.

<sub>By submitting this pull request I agree that all contributions comply with this project's open source license(s).</sub>

### CODE REVIEW

**Comment by Archmonger** [comment]
While the tests may have passed on this PR, the tests are still extremely flakey.

The tests being flakey appears to already be a problem on `main`, and is probably related to `reactpy.testing.DisplayFixture`.

Will need a follow up PR to resolve this.

**Comment by rmorshea** [comment]
Will take a look this weekend.

**Comment by Archmonger** [comment]
Fixed the flakey tests by adding some click delays.

I saw this problem before on `reactpy-django`, where the GitHub CI lags enough to where you can click an element before reactpy finishes re-rendering, which causes a click event to be sent to a non-existent element.

File: src/reactpy_router/components.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+            pathname = urljoin(f"{part_1}/", f"../{part_2}")
+
+        # Resolve relative paths that match `foo/./bar`
+        pathname = pathname.replace("/./", "/")
+
+        set_location(Location(pathname, search))
+
+    attrs["onClick"] = on_click
+
+    return html._(html.a(attrs, *children), html.script(link_js_content.replace("UUID", uuid_string)))
```
Comment (Archmonger) [inline src/reactpy_router/components.py:92]: I know this solution isn't ideal, but I'd rather let this PR move forward with this short-term solution so I can continue tacking the other issues in `reactpy-router`.

I'll need to do some fixes on ReactPy core to allow me to uncomment the client-side approach from `def link()`.

File: src/reactpy_router/hooks.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    The `use_search_params` hook is used to read the query string in the URL \
+    for the current location.
+
+    See `urllib.parse.parse_qs` for info on this hook's parameters."""
+
+    # TODO: In order to match `react-router`, this will need to return a tuple of the search params \
+    # and a function to update them. This is currently not possible without reactpy core having a \
+    # communication layer.
+    return parse_qs(
+        use_location().search[1:],
```
Comment (Archmonger) [inline src/reactpy_router/hooks.py:65]: This might fail if there is no search params

File: src/reactpy_router/resolvers.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            except KeyError as e:
+                raise ValueError(f"Unknown conversion type {param_type!r} in {path!r}") from e
+
+            # Add the string before the match to the pattern
+            pattern += re.escape(path[last_match_end : match.start()])
+
+            # Add the match to the pattern
+            pattern += f"(?P<{name}>{conversion_info['regex']})"
+
+            # Keep a local mapping of parameter names to conversion functions.
```
Comment (Archmonger) [inline src/reactpy_router/resolvers.py:59]:
```suggestion
            # Keep a local mapping of the URL's parameter names to conversion functions.
```

File: src/reactpy_router/components.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+Link = export(
+    module_from_file("reactpy-router", file=Path(__file__).parent / "static" / "bundle.js"),
+    ("Link"),
+)
+"""Client-side portion of link handling"""
+
+link_js_content = (Path(__file__).parent / "static" / "link.js").read_text(encoding="utf-8")
+
+
+def link(attributes: dict[str, Any], *children: Any) -> Component:
```
Comment (Archmonger) [inline src/reactpy_router/components.py:32]: Change `attributes` type to a `TypedDict`

Comment (Archmonger) [reply src/reactpy_router/components.py:32]: Using `TypedDict` with `total=false` was causing a bunch of issues, so I'll skip this for now.

**Pull request merged**